File: asp.websharper.spa/Startup.cs
using asplib.Model.Db;
using asplib.Remoting;
using iselenium;
using Microsoft.AspNetCore;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Net.Http.Headers;
using System;
using WebSharper.AspNetCore;

namespace asp.websharper.spa
{
    public class Startup
    {
        public IConfigurationRoot Configuration { get; }
        public IWebHostEnvironment Environment { get; }
        public static HttpContext HttpContext { get; set; }

        public Startup(IWebHostEnvironment env)
        {
            Environment = env;
            var dom = new ConfigurationBuilder()
                .SetBasePath(env.ContentRootPath)
                .AddJsonFile("appsettings.json", optional: false, reloadOnChange: true)
                .AddEnvironmentVariables()
                .Build();
            Configuration = dom;
            RemotingContext.Set(Environment, Configuration);
            ASP_DBEntities.ConnectionString = Configuration["ASP_DBEntities"];  // globally
        }

        public void ConfigureServices(IServiceCollection services)
        {
            // WebSharper empty
            services.AddDistributedMemoryCache()
                .AddSession(options =>
                {
                    options.IdleTimeout = TimeSpan.FromMinutes(30);
                    options.Cookie.HttpOnly = true;
                })
                .AddLogging()
                .AddSitelet(TestResultSite.Main);
        }

        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                WebSharper.Web.Remoting.DisableCsrfProtection();    // Prevent HTTP 403 errors in GUI tests
                app.UseDeveloperExceptionPage();
            }
            else
            {
                app.UseHttpsRedirection();
                app.UseHsts();
            }
            app.UseMiddleware<ISeleniumMiddleware>()
                .UseSession()
                .UseMiddleware<RequestQuerySessionMiddleware>()
                .UseDefaultFiles()
                .UseStaticFiles()
                .UseWebSharper()
                .Run(context =>
                {
                    HttpContext = context;
                    context.Response.StatusCode = 404;
                    return context.Response.WriteAsync("Page not found");
                });
        }

        public static void Main(string[] args)
        {
            WebHost.CreateDefaultBuilder(args)
                .UseStartup<Startup>()
                .CaptureStartupErrors(true)
                .UseSetting("detailedErrors", "true")
                .Build()
                .Run();
        }
    }
}

File: asp.websharper.spa.fs/Startup.fs
namespace asp.websharper.spa.fs

open asplib.Model.Db
open asplib.Remoting
open iselenium
open System
open Microsoft.AspNetCore.Builder
open Microsoft.AspNetCore.Hosting
open Microsoft.AspNetCore.Http
open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Hosting
open WebSharper.AspNetCore

type Startup private () =
    new (env: IWebHostEnvironment, configuration: IConfiguration) as this =
        Startup() then
        this.Configuration <- configuration
        this.Environment <- env
        let conf =  this.Configuration
        do RemotingContext.Set(env, conf) |> ignore
        do ASP_DBEntities.ConnectionString <- conf.["ASP_DBEntities"]

    member this.ConfigureServices(services: IServiceCollection) =
        services.AddDistributedMemoryCache() |> ignore
        services.AddSession(fun options ->
            options.IdleTimeout <- TimeSpan.FromMinutes(30.0)
            options.Cookie.HttpOnly <- true
        ) |> ignore
        services.AddLogging() |> ignore
        services.AddSitelet(TestResultSite.Main) |> ignore

    member this.Configure(app: IApplicationBuilder, env: IWebHostEnvironment) =

        if (env.IsDevelopment()) then
            app.UseDeveloperExceptionPage() |> ignore
        else
            app.UseHsts() |> ignore

        app.UseMiddleware<IIEMiddleware>() |> ignore
        app.UseDefaultFiles() |> ignore
        app.UseHttpsRedirection() |> ignore
        app.UseStaticFiles() |> ignore
        app.UseSession() |> ignore
        app.UseWebSharper() |> ignore

        app.Run(fun context ->
            context.Response.StatusCode <- 404
            context.Response.WriteAsync("Page not found")) |> ignore

    member val Configuration : IConfiguration = null with get, set
    member val Environment : IWebHostEnvironment = null with get, set

module Program =
    let exitCode = 0

    let CreateHostBuilder args =
        Host.CreateDefaultBuilder(args)
            .ConfigureWebHostDefaults(fun webBuilder ->
                webBuilder.UseStartup<Startup>() |> ignore
                webBuilder.CaptureStartupErrors(true) |> ignore
                webBuilder.UseSetting("detailedErrors", "true") |> ignore
            )

    [<EntryPoint>]
    let main args =
        CreateHostBuilder(args).Build().Run()

        exitCode